From 45b34744ce6b97982aa57994019ef11d8fcd5796 Mon Sep 17 00:00:00 2001
From: Hardik Panchal <hardikx.panchal@intel.com>
Date: Mon, 14 Nov 2022 12:47:22 +0000
Subject: [PATCH] PFR images support in phosphor-software-manager

1. Adds support for handling the PFR images upload and processing.
   Testing:
   Tested PFR image uploads and updates
2. Adds HASH verification on PFR images uploaded for firmware updates
   Tested: tested firmware update with good and bad HASH images.
   A)
   1. Upload the corrupted image for fw update.
   2. Image present in /tmp/images/
   -rw-r--r--    1 root     root      22969344 Jun  3 09:27
   5dea710b-8b85-4065-8af7-3149ada81edf
   3. image deleted from /tmp/images/

   B)
   1. Upload the correct image.
   POST: https://<BMC_IP>/redfish/v1/UpdateService/
         with <BMC_signed_cap> binary file
   2. Image verification is success and proceeds with update.
   {
     "@odata.id": "/redfish/v1/TaskService/Tasks/0",
     "@odata.type": "#Task.v1_4_3.Task",
     "Id": "0",
     "TaskState": "Running",
     "TaskStatus": "OK"
   }

3. Uses report functionality to update error return status for
   redfish updates.
   Tested: Got 400 error with different messages based on failure type
   {
       "error": {
           "@Message.ExtendedInfo": [
               {
                   "@odata.type": "/redfish/v1/$metadata#Message.v1_0_0.Message",
                   "Message": "Invalid file uploaded to /redfish/v1/UpdateService: invalid archive.",
                   "MessageArgs": [
                       "/redfish/v1/UpdateService",
                       "invalid archive"
                   ],
                   "MessageId": "OpenBMC.0.1.0.InvalidFile",
                   "Resolution": "None.",
                   "Severity": "Warning"
               }
           ],
           "code": "OpenBMC.0.1.0.InvalidFile",
           "message": "Invalid file uploaded to /redfish/v1/UpdateService: invalid archive."
       }
   }

   {
       "error": {
           "@Message.ExtendedInfo": [
               {
                   "@odata.type": "/redfish/v1/$metadata#Message.v1_0_0.Message",
                   "Message": "Invalid file uploaded to /redfish/v1/UpdateService: invalid image format.",
                   "MessageArgs": [
                       "/redfish/v1/UpdateService",
                       "invalid image format"
                   ],
                   "MessageId": "OpenBMC.0.1.0.InvalidFile",
                   "Resolution": "None.",
                   "Severity": "Warning"
               }
           ],
           "code": "OpenBMC.0.1.0.InvalidFile",
           "message": "Invalid file uploaded to /redfish/v1/UpdateService: invalid image format."
       }
   }

   {
       "error": {
           "@Message.ExtendedInfo": [
               {
                   "@odata.type": "#Message.v1_0_0.Message",
                   "Message": "The resource /redfish/v1/UpdateService was unable to satisfy the request due to unavailability of resources.",
                   "MessageArgs": [
                       "/redfish/v1/UpdateService"
                   ],
                   "MessageId": "Base.1.4.0.ResourceExhaustion",
                   "Resolution": "Ensure that the resources are available and resubmit the request.",
                   "Severity": "Critical"
               }
           ],
           "code": "Base.1.4.0.ResourceExhaustion",
           "message": "The resource /redfish/v1/UpdateService was unable to satisfy the request due to unavailability of resources."
       }
   }

4. Add support verify the complete fw image by using mtd-util repo's
   pfr_authenticate function.
   Tested.
   1. Upload the corrupted image.
   POST: https://<BMC_IP>/redfish/v1/UpdateService/
         with <Corrupted BMC_signed_cap> binary file
   Response:
   {
     "error": {
       "@Message.ExtendedInfo": [
         {
           "@odata.type": "/redfish/v1/$metadata#Message.v1_0_0.Message",
           "Message": "Invalid file uploaded to /redfish/v1/UpdateService:
           Invalid image format.",
           "MessageArgs": [
             "/redfish/v1/UpdateService",
             "Invalid image format"
           ],
           "MessageId": "OpenBMC.0.1.0.InvalidUpload",
           "Resolution": "None.",
           "Severity": "Warning"
         }
       ],
       "code": "OpenBMC.0.1.0.InvalidUpload",
       "message": "Invalid file uploaded to /redfish/v1/UpdateService:
        Invalid image format."
     }
   }

   2. Upload the correct image.
   POST: https://<BMC_IP>/redfish/v1/UpdateService/
         with <BMC_signed_cap> binary file

   Image verified and firmware updated.
   {
     "@odata.id": "/redfish/v1/TaskService/Tasks/0",
     "@odata.type": "#Task.v1_4_3.Task",
     "Id": "0",
     "TaskState": "Running",
     "TaskStatus": "OK"
   }

   Command:
   GET: https://<BMC_IP>/redfish/v1/Systems/system/LogServices/EventLog/
        Entries

   Response:
   {
         "@odata.id": "/redfish/v1/Systems/system/LogServices/EventLog/
          Entries/1596082187",
         "@odata.type": "#LogEntry.v1_4_0.LogEntry",
         "Created": "2020-07-30T04:09:47+00:00",
         "EntryType": "Event",
         "Id": "1596082187",
         "Message": "BMC firmware update to version 00.72 completed
                     successfully.",
         "MessageArgs": [
           "BMC",
           "00.72"
         ],
         "MessageId": "OpenBMC.0.1.FirmwareUpdateCompleted",
         "Name": "System Event Log Entry",
         "Severity": "OK"
       }

5. Log redfish errors on all pfr image auth failures
   Previous code was doing a 'mtd-util pfr authenticate' prior to manually
   calculating and comparing the hashes. This is incorrect behavior. There
   is no need to manually calculate hashes since that is part of the pfr
   authentication process.

   An unintended side effect of this is that if pfr authenticate fails for
   some reason other than hash compare, the redfish log does not happen.

   Tested: phosphor-version-software-mananger logs a redfish log on rom id
           mismatch as expected.

   Jun 01 05:33:38 intel-obmc phosphor-version-software-manager[430]:
   Firmware image authentication failed
   Jun 01 05:33:38 intel-obmc phosphor-version-software-manager[430]:
   Error verifying uploaded image

Signed-off-by: Vikram Bodireddy <vikram.bodireddy@intel.com>
Signed-off-by: Chalapathi Venkataramashetty <chalapathix.venkataramashetty@intel.com>
Signed-off-by: Vernon Mauery <vernon.mauery@intel.com>
Signed-off-by: Hardik Panchal <hardikx.panchal@intel.com>
Upstream-Status: Pending
---
 bmc/dbus_helpers.hpp       |  30 ++++
 bmc/item_updater.cpp       |   7 +-
 bmc/meson.build            |   8 +-
 meson.options              |   3 +
 7 files changed, 534 insertions(+), 4 deletions(-)
 create mode 100644 bmc/dbus_helpers.hpp
 create mode 100644 bmc/pfr_image_manager.cpp
 create mode 100644 bmc/pfr_image_manager.hpp

diff --git a/bmc/dbus_helpers.hpp b/bmc/dbus_helpers.hpp
new file mode 100644
index 0000000..b9ffa36
--- /dev/null
+++ b/bmc/dbus_helpers.hpp
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "config.h"
+
+#include <sdbusplus/bus.hpp>
+inline bool isFwupdScriptRunning(sdbusplus::bus::bus& bus)
+{
+    using ObjectPath = sdbusplus::message::object_path;
+    // type is ssssssouso
+    using ListUnitsType =
+        std::tuple<std::string, std::string, std::string, std::string,
+                   std::string, std::string, ObjectPath, uint32_t, std::string,
+                   ObjectPath>;
+    auto method = bus.new_method_call(SYSTEMD_BUSNAME, SYSTEMD_PATH,
+                                      SYSTEMD_INTERFACE, "ListUnits");
+
+    auto reply = bus.call(method);
+    std::vector<ListUnitsType> resp;
+    reply.read(resp);
+
+    for (const auto& unit : resp)
+    {
+        if (std::get<0>(unit).find("fwupd@") != std::string::npos &&
+            std::get<3>(unit) != "failed")
+        {
+            return true;
+        }
+    }
+    return false;
+}
diff --git a/bmc/item_updater.cpp b/bmc/item_updater.cpp
index 1173444..df23a2e 100644
--- a/bmc/item_updater.cpp
+++ b/bmc/item_updater.cpp
@@ -68,10 +68,10 @@ void ItemUpdater::createActivation(sdbusplus::message_t& msg)
                     auto value = SVersion::convertVersionPurposeFromString(
                         std::get<std::string>(property.second));
                     if (value == VersionPurpose::BMC ||
-#ifdef HOST_BIOS_UPGRADE
+#if defined(HOST_BIOS_UPGRADE) || defined(PFR_UPDATE)
                         value == VersionPurpose::Host ||
 #endif
-                        value == VersionPurpose::System)
+                        value == VersionPurpose::Other)
                     {
                         purpose = value;
                     }
@@ -570,6 +570,7 @@ void ItemUpdater::deleteAll()
 ItemUpdater::ActivationStatus ItemUpdater::validateSquashFSImage(
     const std::string& filePath)
 {
+#ifndef PFR_UPDATE
     bool valid = true;
 
     // Record the images which are being updated
@@ -587,7 +588,7 @@ ItemUpdater::ActivationStatus ItemUpdater::validateSquashFSImage(
             return ItemUpdater::ActivationStatus::invalid;
         }
     }
-
+#endif
     return ItemUpdater::ActivationStatus::ready;
 }
 
diff --git a/bmc/meson.build b/bmc/meson.build
index ca7e936..913b499 100644
--- a/bmc/meson.build
+++ b/bmc/meson.build
@@ -47,6 +47,7 @@ conf.set('WANT_SIGNATURE_VERIFY', get_option('verify-signature').allowed())
 conf.set('START_UPDATE_DBUS_INTEFACE', get_option('software-update-dbus-interface').allowed())
 
 conf.set('FWUPD_SCRIPT', get_option('fwupd-script').enabled())
+conf.set('PFR_UPDATE', get_option('pfr-update').enabled())
 
 # Configurable variables
 conf.set('ACTIVE_BMC_MAX_ALLOWED', get_option('active-bmc-max-allowed'))

diff --git a/meson.options b/meson.options
index 07105d8..1033145 100644
--- a/meson.options
+++ b/meson.options
@@ -37,6 +37,9 @@ option(
 option('fwupd-script', type: 'feature',
     description: 'Enable fwupd script support.')
 
+option('pfr-update', type: 'feature',
+    description: 'Enable PFR image support.')
+
 # Variables
 option(
     'active-bmc-max-allowed', type: 'integer',
-- 
2.25.1

